File: 00020. Valid Parentheses/00020.c
bool isValid(char * s){
    int i, len = strlen(s), stack_index;
    char stack[10001] = {};
    
    if (len & 1) {  // odd str-length
        return false;
    }
    
    for (i=stack_index=0; i<len; i++) {
        if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
            stack[stack_index++] = s[i];
        } else {
            if (stack_index < 1) {
                return false;
            }
            if (s[i] == ')') {
                if (stack[--stack_index] != '(') {
                    return false;
                }
            } else if (s[i] == ']') {
                if (stack[--stack_index] != '[') {
                    return false;
                }
            } else if (s[i] == '}') {
                if (stack[--stack_index] != '{') {
                    return false;
                }
            }
        }
    }
    
    return stack_index == 0;
}

File: 00020. Valid Parentheses/00020.go
// https://leetcode.com/problems/valid-parentheses

func pop(stack []byte) []byte {
    return stack[:len(stack)-1]
}

func top(stack []byte) byte {
    return stack[len(stack)-1]
}

func isValid(s string) bool {
    if len(s) % 2 != 0 {
        return false
    }

    var stack []byte
    for _, c1 := range s {
        c := byte(c1)
        switch c {
        case '(', '[', '{':
            stack = append(stack, c)
        case ')', ']', '}':
            if len(stack) == 0 {
                return false
            }
            if c == ')' {
                c++
            }
            if c-2 != top(stack) {
                return false
            }
            stack = pop(stack)
        }
    }

    return len(stack) == 0
}
